engine/classes/Elgg/Router.php

```diff
@@ -95,17 +95,24 @@ public function route(\Elgg\Http\Request $request) {
 			$handled = call_user_func($function, $segments, $identifier);
 		}
 
-		$output = ob_get_clean();
+		if ($handled === true) {
+			$output = ob_get_clean();
+		} else if (is_string($handled)) {
+			$output = $handled;
+		}
 
-		$ajax_api = _elgg_services()->ajax;
-		if ($ajax_api->isReady()) {
-			$path = implode('/', $request->getUrlSegments());
-			$ajax_api->respondFromOutput($output, "path:$path");
-			return true;
+		if ($output) {
+			$ajax_api = _elgg_services()->ajax;
+			if ($ajax_api->isReady()) {
+				$path = implode('/', $request->getUrlSegments());
+				$ajax_api->respondFromOutput($output, "path:$path");
+				return true;
+			}
+
+			return _elgg_services()->response->prepareResponse($output)->send();
 		}
 
-		echo $output;
-		return $handled || headers_sent();
+		return headers_sent();
```

Comment:
I thought of allowing instances of Response to be handled here, but wasn't sure if that's such a great idea.